src/Parsers/ExpressionElementParsers.cpp

```diff
@@ -1975,6 +1976,12 @@ bool ParserWithOptionalAlias::parseImpl(Pos & pos, ASTPtr & node, Expected & exp
         if (auto * ast_with_alias = dynamic_cast<ASTWithAlias *>(node.get()))
         {
             tryGetIdentifierNameInto(alias_node, ast_with_alias->alias);
+            // the alias is parametrised and will be resolved later when the query context is known
+            if (!alias_node->children.empty() && alias_node->children.front()->as<ASTQueryParameter>())
```

Comment:
Okay, seems like the specific thing is that we have to call `setAlias` on the parent's node, but it doesn't explain why we need to wrap it with `ASTWithAliasPending`

Reply:
@tavplubix thank you for taking a look.
 `ASTWithAliasPending` is a way to indicate to `ReplaceQueryParameterVisitor` that this AST needs an alias to be set for it when the query parameters are known (that's not the case during parsing iiuc). There are other ways to pass this information to ReplaceQueryParameterVisitor, e.g. set a special value in ASTWithAlias's `alias` itself and then replace it, but it feels much more hacky.
Another obstacle is that ReplaceQueryParameterVisitor needs to distinguish between different types of parametrised identifiers.
For example, it needs to know the difference between `select {parametrized_column_name} from table` and `select known_column_name as {parametrized_alias} from table`.
Both `{parametrized_column_name}` and `known_column_name as {parametrized_alias}` would be represented as Identifier.  The Visitor needs to decide what to do with each of them - replace the value itself or set the alias according to the resolved parameter. IMO introducing ASTWithAliasPending is the cleanest way to distinguish the two use-cases that doesn't involve polluting other types.

Reply:
> Okay, seems like the specific thing is that we have to call setAlias on the parent node, but it doesn't explain why we need to wrap it with ASTWithAliasPending
To answer your question more directly: it's not really about calling setAlias on the parent node, but rather than distinguishing the target of ASTQueryParameter's resolution, like I mentioned above.